### PR TITLE
fix PdfPCell background color ignoring alpha channel of an rgba

### PR DESCRIPTION
**Pull request by sixdouglas**
Fixing the #276 : the PdfPCell background color alpha layer is not taken into account.

### CODE REVIEW

**Comment by sixdouglas** [comment]
I've reworked this PR.
@asturio @mkl-public @razilein : here is a big PR to bring transparency in some of the PDF Elements. 
This work is inspired by the comment of @mkl-public on #276 issue. 
You can try this class to see the result:

`com.lowagie.examples.directcontent.colors.TableCellTransparency`

File: openpdf/src/main/java/com/lowagie/text/pdf/RGBColor.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
+/*
+ * $Id: CMYKColor.java 3427 2008-05-24 18:32:31Z xlv $
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/RGBColor.java]: I don't know it this C&P comment is ok.

File: openpdf/src/main/java/com/lowagie/text/pdf/GrayColor.java [hunk 2]
```diff
@@ -59,18 +59,26 @@
 
     private float gray;
     
-    public static final GrayColor GRAYBLACK = new GrayColor(0f);
-    public static final GrayColor GRAYWHITE = new GrayColor(1f);
+    public static final GrayColor GRAYBLACK = new GrayColor(MAX_FLOAT_COLOR_VALUE);
+    public static final GrayColor GRAYWHITE = new GrayColor(MAX_FLOAT_COLOR_VALUE);
```
[the remark this answers is not in the document]
Comment (sixdouglas) [reply openpdf/src/main/java/com/lowagie/text/pdf/GrayColor.java]: fixed

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 3]
```diff
@@ -112,10 +112,20 @@
         /** The current word spacing */
         protected float wordSpace = 0;
 
+        protected ExtendedColor colorFill = new GrayColor(0);
+
+        protected ExtendedColor colorStroke = new GrayColor(0);
```
[the remark this answers is not in the document]
Comment (sixdouglas) [reply openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: done

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 4]
```diff
@@ -508,6 +530,7 @@ public void setGrayFill(float gray) {
      */
 
     public void resetGrayFill() {
+        saveColorFill(new GrayColor(0));
```
[the remark this answers is not in the document]
Comment (sixdouglas) [reply openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: done

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 5]
```diff
@@ -529,6 +556,7 @@ public void setGrayStroke(float gray) {
      */
 
     public void resetGrayStroke() {
+        saveColorStroke(new GrayColor(0));
```
[the remark this answers is not in the document]
Comment (sixdouglas) [reply openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: done

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 6]
```diff
@@ -579,6 +611,7 @@ public void setRGBColorFillF(float red, float green, float blue) {
      */
 
     public void resetRGBColorFill() {
+        saveColorFill(new GrayColor(0));
```
[the remark this answers is not in the document]
Comment (sixdouglas) [reply openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: that makes sens but it is not the way it is implemented, see next line.
Changed

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 7]
```diff
@@ -607,6 +641,7 @@ public void setRGBColorStrokeF(float red, float green, float blue) {
      */
 
     public void resetRGBColorStroke() {
+        saveColorStroke(new GrayColor(0));
```
[the remark this answers is not in the document]
Comment (sixdouglas) [reply openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: that makes sens but it is not the way it is implemented, see next line.
Changed

**Comment by sixdouglas** [comment]
@travis buildit

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mkl-public) [reply openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: Hhmmmm, considering the method name and the similar `resetCMYKColorFill` implementation, I'd think that a bug.

Considering the JavaDoc comment *"Changes the current color for filling paths to black"* and the calls of that method from within the project, I'd consider it a possible (albeit weird) implementation.

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mkl-public) [reply openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: Hhmmmm, considering the method name and the similar `resetCMYKColorStroke` implementation, I'd think that a bug.

Considering the JavaDoc comment *"Changes the current color for stroking paths to black"* and the calls of that method from within the project, I'd consider it a possible (albeit weird) implementation.

**Comment by sixdouglas** [comment]
I think keeping the gray would be better to avoid changing expected behaviour.

If you look at what iText do in this method [here](https://github.com/itext/itextpdf/blob/develop/itext/src/main/java/com/itextpdf/text/pdf/PdfContentByte.java#L773), they reset to gray either.

**Comment by mkl-public** [comment]
> If you look at what iText do ...

Indeed. And that code even explicitly refers to the Gray methods...

Thus, keep it, but probably also add a hint to the JavaDocs pointing out this discrepancy.

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 8]
```diff
@@ -538,6 +566,9 @@ public void resetGrayStroke() {
      * @param   green   the intensity of green. A value between 0 and 1
      * @param   blue    the intensity of blue. A value between 0 and 1
      */
+    private void HelperRGB(int red, int green, int blue) {
+        HelperRGB((float)(red & 0xFF) / 0xFF, (float)(green & 0xFF) / 0xFF, (float)(blue & 0xFF) / 0xFF);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: There are still some MagicNumbers here 0xFF can the exchanged with MAX_INT_COLOR_VALUE.

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 9]
```diff
@@ -570,16 +601,21 @@ else if (blue > 1.0f)
      */
 
     public void setRGBColorFillF(float red, float green, float blue) {
+        setRGBColorFillF(red, green, blue, 1f);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: 1f is MAX_FLOAT_COLOR_VALUE

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 10]
```diff
@@ -618,6 +655,10 @@ public void resetRGBColorStroke() {
      * @param   yellow  the intensity of yellow. A value between 0 and 1
      * @param   black   the intensity of black. A value between 0 and 1
      */
+    private void HelperCMYK(int cyan, int magenta, int yellow, int black) {
+        HelperCMYK((float)(cyan & 0xFF) / 0xFF, (float)(magenta & 0xFF) / 0xFF, (float)(yellow & 0xFF) / 0xFF, (float)(black & 0xFF) / 0xFF);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: Same here

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 11]
```diff
@@ -654,6 +695,11 @@ else if (black > 1.0f)
      */
 
     public void setCMYKColorFillF(float cyan, float magenta, float yellow, float black) {
+        setCMYKColorFillF(cyan, magenta, yellow, black, 1f);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: Same here

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 12]
```diff
@@ -683,6 +731,10 @@ public void resetCMYKColorFill() {
      */
 
     public void setCMYKColorStrokeF(float cyan, float magenta, float yellow, float black) {
+        setCMYKColorStrokeF(cyan, magenta, yellow, black, 1f);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: Same here

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 13]
```diff
@@ -2164,7 +2208,11 @@ public void setCMYKColorStroke(int cyan, int magenta, int yellow, int black) {
      */
 
     public void setRGBColorFill(int red, int green, int blue) {
-        HelperRGB((float)(red & 0xFF) / 0xFF, (float)(green & 0xFF) / 0xFF, (float)(blue & 0xFF) / 0xFF);
+        setRGBColorFill(red, green, blue, 255);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: MAX_INT_COLOR_VALUE?

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 14]
```diff
@@ -2185,7 +2233,11 @@ public void setRGBColorFill(int red, int green, int blue) {
      */
 
     public void setRGBColorStroke(int red, int green, int blue) {
-        HelperRGB((float)(red & 0xFF) / 0xFF, (float)(green & 0xFF) / 0xFF, (float)(blue & 0xFF) / 0xFF);
+        setRGBColorStroke(red, green, blue, 255);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: Here too?

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 15]
```diff
@@ -2222,10 +2275,18 @@ public void setColorStroke(Color color) {
...
             }
             default:
-                setRGBColorStroke(color.getRed(), color.getGreen(), color.getBlue());
+                setRGBColorStroke(color.getRed(), color.getGreen(), color.getBlue(), color.getAlpha());
         }
     }
 
+    private void saveColorStroke(ExtendedColor extendedColor) {
+        PdfGState gState = new PdfGState();
+        gState.setStrokeOpacity(extendedColor.getAlpha() / 255F);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: Here

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 16]
```diff
@@ -2235,16 +2296,17 @@ public void setColorFill(Color color) {
         int type = ExtendedColor.getType(color);
         switch (type) {
             case ExtendedColor.TYPE_GRAY: {
-                setGrayFill(((GrayColor)color).getGray());
+                GrayColor grayColor = (GrayColor) color;
+                setGrayFill(grayColor.getGray(), color.getAlpha() / 255f);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: Here

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 17]
```diff
@@ -2235,16 +2296,17 @@ public void setColorFill(Color color) {
...
                 break;
             }
             case ExtendedColor.TYPE_CMYK: {
-                CMYKColor cmyk = (CMYKColor)color;
-                setCMYKColorFillF(cmyk.getCyan(), cmyk.getMagenta(), cmyk.getYellow(), cmyk.getBlack());
+                CMYKColor cmyk = (CMYKColor) color;
+                setCMYKColorFillF(cmyk.getCyan(), cmyk.getMagenta(), cmyk.getYellow(), cmyk.getBlack(), cmyk.getAlpha() / 255f);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: Here

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java [hunk 18]
```diff
@@ -2259,17 +2321,26 @@ public void setColorFill(Color color) {
...
             }
             default:
-                setRGBColorFill(color.getRed(), color.getGreen(), color.getBlue());
+                setRGBColorFill(color.getRed(), color.getGreen(), color.getBlue(), color.getAlpha());
         }
     }
 
+    private void saveColorFill(ExtendedColor extendedColor) {
+        PdfGState gState = new PdfGState();
+        gState.setFillOpacity(extendedColor.getAlpha() / 255F);
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfContentByte.java]: Here. using it everywhere :-). Use a static import for the constant if needed.

**Comment by asturio** [comment]
I'm fine with this state now.

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfDocument.java [hunk 19]
```diff
@@ -1481,14 +1483,18 @@ else if (isJustified) {
                         float ascender = chunk.font().getFont().getFontDescriptor(BaseFont.ASCENT, fontSize);
                         float descender = chunk.font().getFont().getFontDescriptor(BaseFont.DESCENT, fontSize);
                         Object[] bgr = (Object[]) chunk.getAttribute(Chunk.BACKGROUND);
-                        graphics.setColorFill((Color)bgr[0]);
+
+                        graphics.setColorFill((ExtendedColor) bgr[0]);
+
```
[the remark this answers is not in the document]
Comment (sixdouglas) [reply openpdf/src/main/java/com/lowagie/text/pdf/PdfDocument.java]: I agree with you. I will fix it

[the remark this answers is not in the document]
Comment (sixdouglas) [reply openpdf/src/main/java/com/lowagie/text/pdf/PdfDocument.java]: Fixed

**Pull request merged**